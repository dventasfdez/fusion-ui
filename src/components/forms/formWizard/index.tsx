export { default as FormWizard } from "./form-wizard";
export { default as FormStep } from "./form-step";
