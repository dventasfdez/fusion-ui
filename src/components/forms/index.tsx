export { default as CheckboxInput } from "./checkbox-input";
export { default as SwitchInput } from "./switch-input";
export { default as RadioInput } from "./radio-input";
export { default as SelectInput } from "./select-input";
export { default as TextArea } from "./text-area";
export { default as HiddenInput } from "./hidden-input";
export { default as DateInput } from "./date-input";
export { default as Form } from "./form";
export { default as TextInput } from "./text-input";
export { default as NumberInput } from "./number-input";
export { default as SelectFilter, Option } from "./select-filter/select-filter";
export { default as FileUploader } from "./file-uploader";
export { default as Wysiwyg } from "./wysiwyg/wysiwyg-base";
export { FormWizard, FormStep } from "./formWizard";
export { serializeForm } from "./utilities/serialize";
