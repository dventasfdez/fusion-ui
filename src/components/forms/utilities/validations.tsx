import { DateTime } from "luxon";

const invalidNameWithHyphenRegExp = /[\x21-\x26\x28-\x2C\x2E-\x2F\x3A-\x40\x5B-\x5D\x7B-\x7D]/;
const validationIsGood = undefined;

export const required = (value: string | string[] | undefined) => {
  if (typeof value === "number") {
    return validationIsGood;
  }
  if (typeof value === "object") {
    const value2: any = value;
    if (value2?.type === "FileSelector") {
      if (!value2?.state?.value) {
        return "validations.required";
      }
    } else if (!value?.length) {
      return "validations.required";
    }
  } else if (typeof value === "string") {
    if (value.replace(/\s/g, "").length > 0) {
      return validationIsGood;
    } else {
      return "validations.required";
    }
  } else if (value !== null && typeof value !== "undefined") {
    return validationIsGood;
  } else {
    return "validations.required";
  }
};
export const isSupportedMimeType = (input: any) => {
  const _accept = input?.props?.accept as string;
  const _current = input?.state?.fileType as string;
  return !_current || _accept?.includes(_current) ? validationIsGood : "validations.file.unsupported";
};

export const maximumFileSize = (input: any, options?: any) => {
  const maxSize = options?.maxSize || input?.props?.maximumSize;
  const fileSize = options?.fileSize || (input?.state?.fileSize as number);
  return !fileSize || fileSize < maxSize ? validationIsGood : "validations.file.tooLarge";
};

export const alphabet = (value: any) => {
  return value && /[\x21-\x26\x28-\x2F\x3A-\x40\x5B-\x5D\x7B-\x7D]/i.test(value) ? "Must be alphabet" : validationIsGood;
};
export const alphanumeric = (value: any) => {
  return value && !/^[A-Z0-9]+$/i.test(value) ? "Must be alphanumeric" : validationIsGood;
};

export const numeric = (value: any) => {
  return value && !/^[0-9]+$/i.test(value) ? "Must be numeric" : validationIsGood;
};

export const uuid = (value: any) => {
  return value && !/^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/i.test(value) ? "Must be a UUID" : validationIsGood;
};

export const email = (value: any) => {
  return value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value) ? "public.descriptionEmail" : validationIsGood;
};
// export const phone = (value: any) => {
//   return value && !regExpHelper.isPhone(value) ? "validations.phone.invalid"  : validationIsGood;
// };
export const _void = (_value: any) => {
  return validationIsGood;
};
export const url = (value: any) => {
  return value && !/^((https|http):\/\/)?(www\.)?(([-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6})|localhost:3001)\b([-a-zA-Z0-9()@:%_+.~#?&/=]*)$/i.test(value)
    ? "Invalid url address"
    : validationIsGood;
};

export const hexacode = (value: any) => {
  return value && !/^#([A-F0-9]{3}|[A-F0-9]{6})$/i.test(value) ? "Invalid hexacode string" : validationIsGood;
};

export const validateAlphabetWithHyphen = (value: string) => {
  return invalidNameWithHyphenRegExp.test(value) ? "validations.invalid.character" : validationIsGood;
};
export const fiscalCode = (value: string) => {
  return value && !/^([A-Za-z]){6,6}([A-Za-z0-9]){2,2}([A-Za-z]){1,1}([A-Za-z0-9]){7,7}$/i.test(value) ? "Please enter a valid fiscal code." : validationIsGood;
};

export const date = (value: any, _context: any, inputComponent: any) => {
  if (!value || Object.prototype.toString.call(value) === "[object Date]") {
    return validationIsGood;
  } else {
    let isDate = null;
    let format = inputComponent && inputComponent.props.dateFormat ? inputComponent.props.dateFormat : null;
    if (value) {
      if (format) {
        isDate = DateTime.fromFormat(value, format).isValid;
      } else {
        isDate = DateTime.fromJSDate(value).isValid;
      }
    }
    if (isDate) {
      return validationIsGood;
    }
    const formatMessage = " (" + format + ")";
    return "Date is not correct" + formatMessage;
  }
};

export const minLength = (minLengthValue: any) => (value: any) => value && value.length < minLengthValue ? `Must be ${minLengthValue} characters or more` : validationIsGood;

export const maxLength = (maxLengthValue: any) => (value: any) => value && value.length > maxLengthValue ? `Must be ${maxLengthValue} characters or less` : validationIsGood;

export const totalLength = (totalLengthValue: any) => (value: any) =>
  value && (value.length > totalLengthValue || value.length < totalLengthValue) ? `Must be ${totalLengthValue} characters` : validationIsGood;

export const min = (minValue: any) => (value: any) => value < minValue ? `Value must be higher than ${minValue}` : validationIsGood;

export const max = (maxValue: any) => (value: any) => value > maxValue ? `Value must be lower than ${maxValue}` : validationIsGood;

export const min1 = (value: any) => (value < 1 ? `Value must be higher than 1` : validationIsGood);

export const passwordsMatch = (value: any, context: any): any => {
  const otherPwdField = context.inputs.find((a: any) => a.props.name === "password");

  return value !== otherPwdField.state.value ? "Passwords don't match" : validationIsGood;
};
export const minValidator = (value: any, allValues: any) => (value > allValues.max ? "Min cannot be higher than Max" : validationIsGood);

export const maxValidator = (value: any, allValues: any) => (value < allValues.min ? "Max cannot be lower than Min" : validationIsGood);

export const minYears = (minYearsValue: any) => (value: any, _context: any, inputComponent: any) => {
  if (value) {
    let _date = null;
    let isValidAge = false;
    let format = inputComponent && inputComponent.props.dateFormat ? inputComponent.props.dateFormat : null;
    // Get date in format
    if (format) {
      _date = DateTime.fromFormat(value, format);
    } else {
      _date = DateTime.fromJSDate(value);
    }
    // Check if is valid
    if (_date.isValid) {
      const years = DateTime.now().diff(_date, "years");
      isValidAge = years >= minYearsValue;
    }
    return !isValidAge ? `Must be ${minYearsValue} age years or more` : validationIsGood;
  }
};

export const minYears16 = minYears(16);
export const minYears18 = minYears(18);
export const minYears21 = minYears(21);

export const mustContainLowerCase = (value: any) => {
  const passwordRegex = new RegExp("^((?=.*[a-z]))");
  return value && !passwordRegex.test(value) ? "Password should contain a lower case letter" : validationIsGood;
};

export const mustContainUpperCase = (value: any) => {
  const passwordRegex = new RegExp("^((?=.*[A-Z]))");
  return value && !passwordRegex.test(value) ? "Password should contain an upper case letter" : validationIsGood;
};

export const mustContainSpecialCharacter = (value: any) => {
  // prettier-ignore
  const passwordRegex = new RegExp("^((?=.*[!@#$%^&*]))");
  return value && !passwordRegex.test(value) ? "Password should contain a Special Character" : validationIsGood;
};

export const isChecked = (input: any, context: any) => {
  if (context && context.inputs && context.inputs.length > 0) {
    return checkIfOneChecked(input, context);
  } else {
    return input && input.state && input.state.checked ? validationIsGood : "validations.required";
  }
};

export const isRadioChecked = (input: any, context: any) => {
  // Maintain this function in order to have separte logic
  // Avoid codesmell
  return isChecked(input, context);
};

function checkIfOneChecked(input: any, context: any) {
  let thereIsOneChecked = false;
  context.inputs.forEach((element: any) => {
    if (!thereIsOneChecked && element.props.name === input.props.name && element.state.checked) {
      thereIsOneChecked = true;
    }
  });
  return thereIsOneChecked ? validationIsGood : "validations.required";
}

export const hasFileDropFiles = (input: any, _context: any) => {
  return input && input.state && input.state.files && input.state.files.length ? validationIsGood : "validations.required";
};
export const hasFileUploaderFiles = (input: any, _context: any) => {
  return input && input.state && input.state.filesValue?.length > 0 ? validationIsGood : "validations.required";
};

export const hasFileSelectorFiles = (input: any) => {
  return input && input.state && input.state.value && input.state.value.length ? validationIsGood : "validations.required";
};

export interface IValidationSummaryItem {
  valid: boolean;
  inputComponent: any;
  name: string;
  value: any;
  errors: string[];
}
export interface IValidationSummary {
  valid: boolean;
  items: IValidationSummaryItem[];
}
/**
 * This function calls behind the sceenes validateInputVerbose with the difference that only returns true or false rather than an object sumarizing the errors.
 * This function goes through all validation functions passed as properties in an input, and uses them to verify the value matches all of them, if the value fails in any of those, the input would be marked as Invalid setting it's stat
 * @returns true or false
 * @param inputComponent - Is the actual input component
 * @param context - Is the form context, where the input is loaded
 */
export async function validateInput(inputComponent: any, context: any, fromFake?: boolean) {
  const validationSumaryItem: IValidationSummaryItem = await validateInputVerbose(inputComponent, context, fromFake);
  return validationSumaryItem.valid;
}

/**
 * This function goes through all validation functions passed as properties in an input, and uses them to verify the value matches all of them, if the value fails in any of those, the input would be marked as Invalid setting it's state
 * @returns an object that will contain the following
 *  IValidationSummaryItem { valid: boolean; inputComponent: any; name: string; value: any; errors: string[]; }
 * @param inputComponent - Is the actual input component
 * @param context - Is the form component, where the input is loaded
 */
export async function validateInputVerbose(inputComponent: any, context: any, fromFake?: boolean) {
  const inputValidations = inputComponent.props.validations || [];
  let validationSumaryItem: IValidationSummaryItem = {
    valid: true,
    inputComponent: inputComponent,
    name: inputComponent?.props?.name,
    value: inputComponent?.state?.value,
    errors: [],
  };
  if (inputComponent.props.visible !== false && (inputValidations?.length || inputComponent.state.requireByRelated) && !inputComponent.state.disabled) {
    let arrayOfErrors: string[] = [];
    if (inputComponent._isMounted) {
      inputComponent.setState({ loading: true });
    }
    // Required related
    if (inputComponent.state.requireByRelated) {
      const errorRelatedRequired = required(inputComponent.state.value);
      if (typeof errorRelatedRequired !== "undefined") arrayOfErrors.push("validations.required.related");
    }
    for await (const validationFunction of inputValidations) {
      if (typeof validationFunction === "function") {
        const type = inputComponent.state.type;
        let errorMessage: string | undefined = "";
        switch (type) {
          case "checkbox":
            if (validationFunction === required) errorMessage = isChecked(inputComponent, context);
            else errorMessage = await validationFunction(inputComponent.state.value, context, inputComponent, fromFake);
            break;
          case "radio":
            if (validationFunction === required) errorMessage = isRadioChecked(inputComponent, context);
            else errorMessage = await validationFunction(inputComponent.state.value, context, inputComponent, fromFake);
            break;
          case "FileDrop":
            errorMessage = hasFileDropFiles(inputComponent, context);
            break;
          case "FileUploader":
            errorMessage = hasFileUploaderFiles(inputComponent, context);
            break;
          case "FileSelector": //add required validator if content is mandatory
            errorMessage = validationFunction(inputComponent);
            break;
          default:
            errorMessage = await validationFunction(inputComponent.state.value, context, inputComponent, fromFake);
        }

        if (typeof errorMessage !== "undefined") {
          arrayOfErrors.push(errorMessage);
        }
        validationSumaryItem.errors = arrayOfErrors;
      }
    }
    if (arrayOfErrors && arrayOfErrors.length) {
      if (inputComponent._isMounted) {
        inputComponent.setState({
          errors: arrayOfErrors,
          isValid: false,
          loading: false,
        });
      }
      validationSumaryItem.valid = false;
      return validationSumaryItem;
    } else {
      if (inputComponent._isMounted) {
        inputComponent.setState({ errors: null, isValid: true, loading: false });
      }
      return validationSumaryItem;
    }
    // If not validation is valid
  } else {
    if (inputComponent._isMounted) {
      inputComponent.setState({ isValid: true });
    }
    return validationSumaryItem;
  }
}
