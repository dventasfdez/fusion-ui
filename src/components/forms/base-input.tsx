import React, { SyntheticEvent, createRef } from "react";

import { FormContext } from "./form";
import { validateInput } from "./utilities/validations";
import Tooltip from "../tooltip/tooltip";

export interface IProps {
  id: string;
  name: string;
  type?: string;
  className?: string;
  value?: string | string[];
  disabled?: boolean;
  disableRelated?: string[];
  label?: string | Function;
  errors?: string[];
  validations?: Function[];
  validateOnChange?: boolean;
  readOnly?: boolean;
  loading?: boolean;
  placeholder?: string;
  onChange?: Function;
  autoComplete?: string;
  tooltip?: string;
  icon?: string;
  [others: string]: any;
  //previous line avoids having to write specific component properties
  //rows?: number; //textarea only
  //cols?: number; //textarea only
}

export interface IState {
  errors?: string[];
  isValid?: boolean;
  value?: string | string[];
  validateOnChange?: boolean;
  parentFormOnChange?: Function;
  readOnly?: boolean;
  loading?: boolean;
  disabled?: boolean;
  disableRelated?: string | string[];
  requireByRelated?: boolean;
  requireRelated?: string | string[];
  [others: string]: any;
}

class BaseInput extends React.Component<IProps, IState> {
  type: string = "";
  _isMounted: boolean = false;
  value: string | string[] | undefined = "";
  validationClass: string | undefined = "";
  validationNames: string[] | undefined = [];
  loadingClass: string | undefined = "";
  inputRef: any = createRef();
  tooltipRef = createRef<HTMLSpanElement>();

  constructor(props: any) {
    super(props);

    this.state = {
      errors: this.props.errors,
      isValid: undefined,
      value: this.props.value !== null && this.props.value !== undefined ? this.props.value : "",
      validateOnChange: this.props.validateOnChange === false ? this.props.validateOnChange : true,
      readOnly: this.props.readOnly,
      loading: this.props.loading,
      disabled: this.props.disabled,
      disableRelated: this.props.disableRelated,
      requireByRelated: false,
      requireRelated: this.props.requireRelated,
      children: this.props.children,
      options: this.props.options,
    };
    this._isMounted = false;

    this.type = "";
    this.change = this.change.bind(this);
  }

  componentDidMount() {
    this._isMounted = true;
    if (this.context && typeof (this.context as any).addInputToContext === "function") {
      (this.context as any).addInputToContext(this);
    }
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  componentDidUpdate(prevProps: IProps, prevState: IState) {
    if (this._isMounted) {
      if (this.props.value !== prevProps.value) {
        this.setState({ value: this.props.value });
      }
      if (this.props.readOnly !== prevProps.readOnly) {
        this.setState({ readOnly: this.props.readOnly });
      }
      if (this.props.loading !== prevProps.loading) {
        this.setState({ loading: this.props.loading });
      }
      if (this.props.tooltip !== prevProps.tooltip) {
        this.setState({ tooltip: this.props.tooltip });
      }
      if (this.props.disabled !== prevProps.disabled) {
        this.setState({ disabled: this.props.disabled });
      }
      if (this.props.disableRelated !== prevProps.disableRelated) {
        this.setState({ disableRelated: this.props.disableRelated });
      }
      if (this.props.requireRelated !== prevProps.requireRelated) {
        this.setState({ requireRelated: this.props.requireRelated });
      }

      if (this.props.children !== prevProps.children) {
        this.setState({ children: this.props.children });
      }
      if (this.props.options !== prevProps.options) {
        this.setState({ options: this.props.options });
      }
      try {
        if (JSON.stringify(prevState) !== JSON.stringify(this.state)) {
          //just for you sonar =),  to avoid wrong code smell for prevState, needed for checkbox input
        }
      } catch (error) {}
    }
  }

  onChange = async (e: SyntheticEvent) => {
    const target: any = e.target as HTMLInputElement;
    let newValue = target.value;
    if (this.type === "SelectInput" && this.props.multiple) {
      const options = target.options;
      newValue = [];
      for (let i = 0, l = options.length; i < l; i++) {
        if (options[i].selected) {
          newValue.push(options[i].value);
        }
      }
    }
    e.persist();
    this.processChange(newValue, e);
  };

  change = (value: string | string[], forceUserEvent: boolean | undefined = false) => {
    const isFakeEvent = forceUserEvent ? false : true;
    const fakeEvent = {
      target: { ...this.inputRef, value: value, fakeEvent: isFakeEvent },
    };

    this.processChange(value, fakeEvent);
  };

  //this is to be overrided by inheritance inputs and set rules to decide whether the value can be set or not: ie In select dropdowns only if the value exist as option
  valueCanBeSet = (value: string | string[], e: SyntheticEvent | any = undefined) => {
    return true;
  };
  // To override it
  changeHiddenInput = (value: string | string[]) => {};

  processChange = (value: string | string[], e: SyntheticEvent | any = undefined) => {
    if (this.valueCanBeSet(value, e)) {
      this.setState({ value: value }, async () => {
        if (this.state.validateOnChange && !e?.target?.fakeEvent) {
          validateInput(this, this.context);
        }
        if (e && this.props.onChange) {
          this.props.onChange(e);
        }

        if (e && !e?.target?.fakeEvent && typeof this.state.parentFormOnChange === "function") {
          this.state.parentFormOnChange(e);
        }
        if (this.type === "SelectInput") this.changeHiddenInput(value);
      });
    }
  };

  renderErrorMessage = () => {
    if (this.state.errors) {
      return (
        <div className="input-helper-text">
          <span className="material-icons">info</span>
          {this.state.errors[0]}
        </div>
      );
    }
  };

  processCSSClasses = () => {
    this.value = this.state.value ? this.state.value : "";
    this.validationClass = "";
    if (typeof this.state.isValid !== "undefined" && this.state.isValid !== null) {
      if (!this.state.isValid || this.value !== "") {
        // Show only when error or is valid but not empty
        this.validationClass = this.state.isValid ? "valid" : "error";
      }
    }
    // if (this.props.visible !== false && ((this.props.validations && this.props.validations.indexOf(required) !== -1) || this.state.requireByRelated)) {
    //   this.validationClass += " validation-required";
    // }
    this.loadingClass = this.state && this.state.loading ? "loading" : "";
    if (this.state.loading) {
      this.loadingClass = "loading";
    }
  };

  getValidations = () => {
    if (this.props.validations && this.props.validations.length) this.validationNames = this.props.validations?.map((validation) => validation.name);
  };

  renderTooltip(parentRef?: any) {
    if (this.props.tooltip) {
      return (
        <Tooltip parentRef={parentRef} renderAsPortal>
          {this.props.tooltip}
        </Tooltip>
      );
    }
  }

  renderTooltipIcon() {
    return this.props.tooltip ? (
      <span className="material-icons info-tooltip ml1" ref={this.tooltipRef}>
        info
        {this.renderTooltip(this.tooltipRef)}
      </span>
    ) : null;
  }

  renderInputLabel() {
    const isLabelNextToInput = this.type === "CheckboxInput" || this.type === "RadioInput" || this.type === "SwitchInput";
    let labelClassName = isLabelNextToInput ? "" : "caption";
    labelClassName = this.state?.isValid === false && isLabelNextToInput ? labelClassName + " error" : labelClassName;
    this.getValidations();
    if (this.props.label) {
      return (
        <label className={labelClassName} htmlFor={this.props.id}>
          {this.validationNames?.includes("required") && <small className="required">*</small>}
          {typeof this.props.label === "function" ? this.props.label() : this.props.label}
          {this.renderTooltipIcon()}
        </label>
      );
    }
  }

  render() {
    return (
      <></> //this get overrided by inheritances
    );
  }
}

BaseInput.contextType = FormContext;
export default BaseInput;
