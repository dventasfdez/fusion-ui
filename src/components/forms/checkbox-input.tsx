import React, { createRef } from "react";
import { FormContext } from "./form";
import { validateInput } from "./utilities/validations";
import BaseInput, { IState } from "./base-input";

class CheckboxInput extends BaseInput {
  checkboxRef = createRef<HTMLInputElement>();

  tooltipRef = createRef<HTMLSpanElement>();
  type = "CheckboxInput";
  state: Readonly<IState> = {
    disabled: this.props.disabled,
    errors: this.props.errors,
    isValid: undefined,
    type: "checkbox",
    value: this.props.value ? this.props.value : "value",
    checked: this.props.checked,
    validateOnChange: this.props.validateOnChange === false ? this.props.validateOnChange : true,
  };

  componentDidMount() {
    this._isMounted = true;
    if (this.context && typeof (this.context as any).addInputToContext === "function") {
      (this.context as any).addInputToContext(this);

      this.updateParentGroup();
    }
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  componentDidUpdate(prevProps: any, prevState: any) {
    if (JSON.stringify(prevProps) !== JSON.stringify(this.props) || JSON.stringify(prevState) !== JSON.stringify(this.state)) {
      if (JSON.stringify(prevProps) !== JSON.stringify(this.props)) {
        if (typeof this.props.checked !== "undefined" && this.props.checked !== prevProps.checked) {
          this.setState({ checked: this.props.checked });
        }
        if (this.props.defaultValue !== prevProps.defaultValue) {
          this.setState({ value: this.props.defaultValue });
        }
        if (this.props.disabled !== prevProps.disabled) {
          this.setState({ disabled: this.props.disabled });
        }
      }

      if (JSON.stringify(prevState) !== JSON.stringify(this.state)) {
        this.updateParentGroup();
      }
    }
  }

  updateParentGroup = () => {
    if (this.props.changeParentGroup && typeof this.props.changeParentGroup === "function") {
      let isChecked = typeof this.state.checked !== "undefined" ? this.state.checked : false;
      this.props.changeParentGroup(this, isChecked);
    }
  };

  onChange = async (SyntheticEvent: any) => {
    const self = this;
    SyntheticEvent.persist();
    this.setState({ checked: !this.state.checked }, () => {
      validateInput(self, this.context);
      //validate siblings
      if (self.context && (self.context as any).inputs && (self.context as any).inputs.length) {
        //checkboxes inside form
        const siblings = (self.context as any).inputs.filter((el: any) => el !== self && el.type === "CheckboxInput" && el.props.name === self.props.name);
        siblings.forEach((element: any) => {
          validateInput(element, self.context);
        });
      }
      this.updateParentGroup();
    });
    if (typeof this.props.onChange === "function") {
      this.props.onChange(SyntheticEvent);
    }

    if (!this.props.disableFormPropagation && SyntheticEvent && !SyntheticEvent?.target?.fakeEvent && typeof this.state.parentFormOnChange === "function") {
      this.state.parentFormOnChange(SyntheticEvent);
    }
  };

  render() {
    this.processCSSClasses();

    const checked = typeof this.state.checked !== "undefined" && this.state.checked !== null ? this.state.checked : false;
    return (
      <div className={`checkbox-container${this.state.disabled ? "_disabled" : ""} ${this.validationClass} ${this.loadingClass} ${this.props.className || ""}`}>
        <input
          data-testid={`checkbox-${this.props.name}`}
          ref={this.checkboxRef}
          disabled={this.state.disabled}
          name={this.props.name}
          value={this.props.value}
          id={this.props.id}
          type="checkbox"
          checked={checked}
          onChange={this.onChange}
          className={this.props.intermediate ? "intermediate" : ""}
        />
        {this.renderInputLabel()}
      </div>
    );
  }
}

CheckboxInput.contextType = FormContext;
export default CheckboxInput;
