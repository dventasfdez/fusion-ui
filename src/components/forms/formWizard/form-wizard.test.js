import React from "react";
import renderer from "react-test-renderer";
import ReactTestUtils, { act } from "react-dom/test-utils";
import ReactDOM from "react-dom";
import { FormWizard, FormStep } from ".";
import { TextInput, CheckboxInput, RadioInput, TextArea } from "../forms";
import { required, email } from "../forms/utilities/validations";
import { cleanup } from "@testing-library/react";

let formWizardRef = undefined;
const submitWizard = (data) => {
  return data;
};

afterEach(cleanup);

class FormWizardExample extends React.Component {
  render() {
    formWizardRef = React.createRef();
    return (
      <FormWizard ref={formWizardRef} onSubmit={(data) => submitWizard(data)} hideSubmitButton={false}>
        <FormStep>
          <div className="p10">
            <p className="h4">step 1</p>
            <TextInput label={"Name"} name="name" type="text" key={1} validations={[required]} />
            <TextInput label={"Surname"} name="surname" type="text" key={2} validations={[]} />
          </div>
        </FormStep>
        <FormStep>
          <div className="p10">
            <p className="h4">step 2</p>
            <TextArea validations={[required]} placeholder="Default text" key={3} name="myTextArea" label="Default Textarea" validateBeforeSubmit={false} />
            <TextInput label={"Email"} name="email" type="email" key={4} validations={[required, email]} />
          </div>
        </FormStep>
        <FormStep>
          <div className="p10">
            <p className="h4">step 3</p>
            <TextInput label={"Address line 1"} name="addr1" type="text" validations={[required]} />
            <TextInput label={"Address line 1"} name="addr2" type="text" validations={[required]} />
            <TextInput label={"Address line 1"} name="addr3" type="text" validations={[required]} />
          </div>
        </FormStep>
        <FormStep>
          <div className="p10">
            <p className="h4">step 4</p>
            <div className="pl10">
              <CheckboxInput name="timeSensitive" type="checkbox" label="Once a week" value={"1"} />
            </div>
            <div className="pl10">
              <CheckboxInput name="temperatureSensitive" type="checkbox" label="Potatoes" value={"1"} />
            </div>
            <div className="pl10">
              <CheckboxInput name="critical" type="checkbox" label="Critical" value={"1"} />
            </div>
            <div className="pl10">
              <RadioInput validations={[required]} name="myRadio" validateBeforeSubmit={false} label="radio 1" value="radio1" />
              <RadioInput validations={[required]} name="myRadio" validateBeforeSubmit={false} label="radio 2" value="radio2" />
              <RadioInput className="mb20" validations={[required]} name="myRadio" validateBeforeSubmit={false} label="radio 3" value="radio3" />
            </div>
          </div>
        </FormStep>
        <FormStep>
          <div className="p10 mb-4">
            <h3>Thanks for your interest</h3>
            <button type="submit" className="btn btn-primary ">
              Submit my data
            </button>
          </div>
        </FormStep>
      </FormWizard>
    );
  }
}

test("Should render", () => {
  const component = renderer.create(<FormWizardExample />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Should NOT move to next step if form is not valid", async () => {
  let container = document.createElement("div");
  document.body.appendChild(container);
  ReactDOM.render(<FormWizardExample />, container);
  const _formWizardRef = formWizardRef;
  try {
    await _formWizardRef.current.formRef.current.submit();
    await new Promise((r) => setTimeout(r, 1)); //wait for the updatestates related to submit
    expect(container.querySelectorAll("textarea[name='myTextArea']").length).toBe(0);
    expect(container.querySelectorAll("input[name='email']").length).toBe(0);
  } catch (ex) {}
});

test("Should move to next step", async () => {
  let container = document.createElement("div");
  document.body.appendChild(container);
  ReactDOM.render(<FormWizardExample />, container);
  ReactTestUtils.Simulate.change(container.querySelector("*[name='name']"), {
    target: { value: "NameValue" },
  });
  ReactTestUtils.Simulate.change(container.querySelector("*[name='surname']"), {
    target: { value: "SurnameValue" },
  });
  const _formWizardRef = formWizardRef;
  act(async () => {
    _formWizardRef.current.formRef.current.submit();
  }).then(() => {
    expect(container.querySelectorAll("textarea[name='myTextArea']").length).toBe(1);
    expect(container.querySelectorAll("input[name='email']").length).toBe(1);
  });
});

// test("Should move to 3rd step", async () => {
//   let container = document.createElement("div");
//   document.body.appendChild(container);
//   ReactDOM.render(<FormWizardExample />, container);
//   ReactTestUtils.Simulate.change(container.querySelector("*[name='name']"), {
//     target: { value: "NameValue" },
//   });
//   ReactTestUtils.Simulate.change(container.querySelector("*[name='surname']"), {
//     target: { value: "SurnameValue" },
//   });
//   const _formWizardRef = formWizardRef;

//   act(async () => {
//     _formWizardRef.current.formRef.current.submit();
//   }).then(() => {
//     expect(container.querySelectorAll("textarea[name='myTextArea']").length).toBe(1);
//     ReactTestUtils.Simulate.change(container.querySelector("textarea[name='myTextArea']"), {
//       target: { value: "myTextAreaValue" },
//     });
//     ReactTestUtils.Simulate.change(container.querySelector("input[name='email']"), {
//       target: { value: "email@value.com" },
//     });
//     act(async () => {
//       _formWizardRef.current.formRef.current.submit();
//     }).then(() => {
//       setImmediate(() => {
//         expect(container.querySelectorAll("input[name='addr1']").length).toBe(1);
//         expect(container.querySelectorAll("input[name='addr2']").length).toBe(1);
//         expect(container.querySelectorAll("input[name='addr3']").length).toBe(1);
//       });
//     });
//   });
// });

test("State should match snapshot after navigating ", async () => {
  let container = document.createElement("div");
  document.body.appendChild(container);
  ReactDOM.render(<FormWizardExample />, container);
  ReactTestUtils.Simulate.change(container.querySelector("*[name='name']"), {
    target: { value: "NameValue" },
  });
  ReactTestUtils.Simulate.change(container.querySelector("*[name='surname']"), {
    target: { value: "SurnameValue" },
  });
  const _formWizardRef = formWizardRef;
  act(async () => {
    _formWizardRef.current.formRef.current.submit();
  }).then(async () => {
    await new Promise((r) => setTimeout(r, 1)); //wait for the updatestates related to submit
    let state = JSON.stringify(_formWizardRef.current.state.values);
    expect(state).toEqual('{"name":"NameValue","surname":"SurnameValue"}');
  });
});
