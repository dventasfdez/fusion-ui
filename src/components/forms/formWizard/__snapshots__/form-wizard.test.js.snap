// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render 1`] = `
<form
  className="-step step-1"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="row d-flex form-wizard-btn pb-5 top-buttons-group"
  >
    <button
      className="btn btn-primary fr"
      type="submit"
    >
      Next
    </button>
    <div
      className="clearfix "
    />
  </div>
  <div
    className="form-wizard-content"
  >
    <div
      className="p10"
    >
      <p
        className="h4"
      >
        step 1
      </p>
      <div
        className="input-wrapper input-wrapper  validation-required  "
      >
        <label
          className="caption"
        >
          <small
            className="required"
          >
            *
          </small>
          <span
            className="label-text"
          >
            Name
          </span>
        </label>
        <div
          className="input-container"
        >
          <input
            name="name"
            onChange={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="input-wrapper input-wrapper   "
      >
        <label
          className="caption"
        >
          
          <span
            className="label-text"
          >
            Surname
          </span>
        </label>
        <div
          className="input-container"
        >
          <input
            name="surname"
            onChange={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="row d-flex form-wizard-btn pb-5 bottom-buttons-group"
  >
    <button
      className="btn btn-primary fr"
      type="submit"
    >
      Next
    </button>
    <div
      className="clearfix "
    />
  </div>
</form>
`;
